vault/src/metrics.rs

```diff
@@ -0,0 +1,527 @@
+                    .count() as f64,
+            );
+            vault.metrics.redeems.completed_count.set(
+                relevant_redeems
+                    .iter()
+                    .filter(|status| matches!(status, RedeemRequestStatus::Completed))
+                    .count() as f64,
+            );
+            vault.metrics.redeems.expired_count.set(
+                relevant_redeems
+                    .iter()
+                    .filter(|status| {
+                        matches!(
+                            status,
+                            RedeemRequestStatus::Reimbursed(_) | RedeemRequestStatus::Retried
+                        )
+                    })
+                    .count() as f64,
+            );
+        }
+    }
+}
+
+pub async fn monitor_bridge_metrics<B: BitcoinCoreApi + Clone + Send + Sync>(
+    parachain_rpc: InterBtcParachain,
+    vault_id_manager: VaultIdManager<B>,
+) -> Result<(), ServiceError> {
+    let parachain_rpc = &parachain_rpc;
+    let vault_id_manager = &vault_id_manager;
+    parachain_rpc
+        .on_event::<FeedValuesEvent, _, _, _>(
+            |event| async move {
+                let updated_currencies = event.values.iter().filter_map(|(key, _value)| match key {
+                    OracleKey::ExchangeRate(currency_id) => Some(currency_id),
+                    _ => None,
+                });
+                let vaults = vault_id_manager.get_entries().await;
+                for currency_id in updated_currencies {
+                    for vault in vaults
+                        .iter()
+                        .filter(|vault| &vault.vault_id.collateral_currency() == currency_id)
+                    {
+                        if let Err(err) =
+                            update_bridge_metrics(parachain_rpc.clone(), vault.vault_id.clone(), vault.metrics.clone())
+                                .await
+                        {
+                            tracing::info!("Failed to update prometheus bridge metrics: {}", err);
+                        }
+                    }
+                }
+            },
+            |error| tracing::error!("Error reading SetExchangeRate event: {}", error.to_string()),
+        )
+        .await?;
+    Ok(())
+}
+
+pub async fn poll_metrics<B: BitcoinCoreApi + Clone + Send + Sync>(
+    parachain_rpc: InterBtcParachain,
+    vault_id_manager: VaultIdManager<B>,
+) -> Result<(), ServiceError> {
+    let parachain_rpc = &parachain_rpc;
+    let vault_id_manager = &vault_id_manager;
+
+    loop {
+        publish_kint_balance(&parachain_rpc).await;
+        publish_issue_count(&parachain_rpc.clone(), &vault_id_manager).await;
+        publish_redeem_count(&parachain_rpc.clone(), &vault_id_manager).await;
+
+        for vault in vault_id_manager.get_entries().await {
+            publish_utxo_count(&vault).await;
+        }
+
+        sleep(SLEEP_DURATION).await;
+    }
```

Comment:
True. The reason I initially went for this approach is because I wanted to have 2 different counters for expired & cancelled, and we don't have parachain events for redeem expiration. In the end I forgot to actually implement it this way though. Anyway, I would propose to leave it like this for now and leave it as a future optimization